plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,555 @@
+            return StaticCredentialsProvider.create(AwsBasicCredentials.create(config.getS3AwsAccessKey(), config.getS3AwsSecretKey()));
+        }
+        return DefaultCredentialsProvider.create();
+    }
+
+    private S3Client createS3Client(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3ClientBuilder clientBuilder = S3Client.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        region.ifPresent(clientBuilder::region);
+        endpoint.ifPresent(s3Endpoint -> clientBuilder.endpointOverride(URI.create(s3Endpoint)));
+
+        return clientBuilder.build();
+    }
+
+    private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3AsyncClientBuilder clientBuilder = S3AsyncClient.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        region.ifPresent(clientBuilder::region);
+        endpoint.ifPresent(s3Endpoint -> clientBuilder.endpointOverride(URI.create(s3Endpoint)));
+
+        return clientBuilder.build();
+    }
+
+    @NotThreadSafe
+    private static class S3ExchangeStorageWriter
+            implements ExchangeStorageWriter
+    {
+        private final S3AsyncClient s3AsyncClient;
+        private final String bucketName;
+        private final String key;
+        private final int partSize;
+        private final Optional<SecretKey> secretKey;
+
+        private int currentPartNumber;
+        private CompletableFuture<Void> directUploadFuture;
+        private CompletableFuture<CreateMultipartUploadResponse> multiPartUploadIdFuture;
+        private final List<CompletableFuture<CompletedPart>> multiPartUploadFutures = new ArrayList<>();
+        private volatile boolean closed;
+
+        public S3ExchangeStorageWriter(S3AsyncClient s3AsyncClient, String bucketName, String key, int partSize, Optional<SecretKey> secretKey)
+        {
+            this.s3AsyncClient = requireNonNull(s3AsyncClient, "s3AsyncClient is null");
+            this.bucketName = requireNonNull(bucketName, "bucketName is null");
+            this.key = requireNonNull(key, "key is null");
+            this.partSize = partSize;
+            this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        }
+
+        @Override
+        public CompletableFuture<Void> write(Slice slice)
+        {
+            checkState(directUploadFuture == null, "Direct upload already started");
+            if (closed) {
+                // Simply ignore after writer is closed
+                return completedFuture(null);
+            }
+
+            // skip multipart upload if there would only be one part
+            if (slice.length() < partSize && multiPartUploadIdFuture == null) {
```

Comment:
Hmm I don't feel it's error-prone?

let consider random sequences:
`fullbuffer`, `fullbuffer` -> we will go through the multipart upload path.
`fullbuffer`, `not-fullbuffer` -> we will go through the multipart upload path.
`not-fullbuffer`, `not-fullbuffer` -> we will direct upload first, and on the second one we will throw IllegalStateException saying "Direct upload already started"
`not-fullbuffer`, `fullbuffer`-> we will direct upload first, and on the second one we will throw IllegalStateException saying "Direct upload already started"

It will be confusing to add the boolean flag to the `write` function.